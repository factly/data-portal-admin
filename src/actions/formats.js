import axios from 'axios';
import {
  ADD_FORMAT,
  ADD_FORMATS,
  SET_FORMAT_LOADING,
  SET_FORMAT_REQUEST,
  SET_FORMAT_IDS,
  RESET_FORMAT,
  FORMAT_API,
} from '../constants/formats';
import { getIds, buildObjectOfItems } from '../utils/objects';

export const loadFormats = (query) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    const response = await axios({
      url:FORMAT_API,
      method: 'get',
      params: query
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { ...query, ids: currentPageIds };
    dispatch(setFormatRequest(currentReq, total));
    dispatch(addFormats(nodes));
    dispatch(setFormatIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createFormat = (data) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    await axios({
      url: FORMAT_API,
      method: 'post',
      data: data,
    });

    dispatch(resetFormat());
    dispatch(setLoading(false));
  };
};

export const updateFormat = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${FORMAT_API}/${id}`;

    dispatch(setLoading(true));

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addFormat(response.data));
    dispatch(setLoading(false));
  };
};

export const deleteFormat = (id) => {
  return async (dispatch, getState) => {
    let url = `${FORMAT_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetFormat());
    dispatch(setLoading(false));
  };
};

export const getFormat = (id) => {
  return async (dispatch, getState) => {
    const {
      formats: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${FORMAT_API}/${id}`,
      method: 'get',
    });
    dispatch(addFormat(response.data));

    dispatch(setLoading(false));
  };
};

export const setLoading = (loading) => {
  return {
    type: SET_FORMAT_LOADING,
    payload: { loading },
  };
};

export const addFormat = (format) => {
  return {
    type: ADD_FORMAT,
    payload: { format },
  };
};

export const addFormats = (formats) => {
  return {
    type: ADD_FORMATS,
    payload: {
      formats: buildObjectOfItems(formats),
    },
  };
};

export const setFormatRequest = (req, total) => {
  return {
    type: SET_FORMAT_REQUEST,
    payload: { req, total },
  };
};

export const setFormatIds = (ids) => {
  return {
    type: SET_FORMAT_IDS,
    payload: { ids },
  };
};

export const resetFormat = () => {
  return {
    type: RESET_FORMAT,
  };
};
