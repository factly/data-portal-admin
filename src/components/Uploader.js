import React from 'react';
import Uppy from '@uppy/core';
import AwsS3 from '@uppy/aws-s3';
import GoogleDrive from '@uppy/google-drive';
import Url from '@uppy/url';
import { DashboardModal } from '@uppy/react';
import '@uppy/core/dist/style.css';
import '@uppy/dashboard/dist/style.css';
import '@uppy/url/dist/style.css';

import { Button } from 'antd';

const Uploader = ({ onUploadSuccess, metaFields }) => {
  const [show, setShow] = React.useState(false);

  const uppy = Uppy({
    id: 'uppy-media',
    meta: { type: 'avatar' },
    restrictions: {
      maxNumberOfFiles: 1,
    },
    autoProceed: false,
    onBeforeUpload: (files) => {
      const updatedFiles = {};
      Object.keys(files)
        .filter((fileId) => files[fileId].extension)
        .forEach((fileId) => {
          updatedFiles[fileId] = files[fileId];
        });
      if (Object.keys(updatedFiles).length === 0) {
        uppy.info('File must have a valid file extension');
        return false;
      }

      return updatedFiles;
    },
  })
    .use(AwsS3, { companionUrl: 'http://localhost:3020' })
    .use(Url, { companionUrl: 'http://localhost:3020' })
    .use(GoogleDrive, { companionUrl: 'http://localhost:3020' })
    .on('complete', (result) => {
      onUploadSuccess(result.successful);
      setShow(false);
    });

  return (
    <div>
      <Button onClick={() => setShow(true)}>Upload</Button>
      <DashboardModal
        uppy={uppy}
        closeModalOnClickOutside
        open={show}
        onRequestClose={() => setShow(false)}
        plugins={['GoogleDrive', 'Url']}
        metaFields={metaFields}
      />
    </div>
  );
};

export default Uploader;
