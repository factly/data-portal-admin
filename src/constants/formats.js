export const ADD_FORMAT = 'ADD_FORMAT';
export const ADD_FORMATS = 'ADD_FORMATS';
export const SET_FORMAT_LOADING = 'SET_FORMAT_LOADING';
export const SET_FORMAT_REQUEST = 'SET_FORMAT_REQUEST';
export const SET_FORMAT_IDS = 'SET_FORMAT_IDS';
export const RESET_FORMAT = 'RESET_FORMAT';

export const FORMAT_API = `/formats`;
