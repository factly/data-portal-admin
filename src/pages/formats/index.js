import React from 'react';
import FormatsList from './components/FormatsList';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

const Formats = () => {
  return (
    <>
      <Link to={'/formats/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Format
        </Button>
      </Link>
      <FormatsList />
    </>
  );
};

export default Formats;
