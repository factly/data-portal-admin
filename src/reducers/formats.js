import produce from 'immer';
import {
  ADD_FORMAT,
  ADD_FORMATS,
  SET_FORMAT_LOADING,
  SET_FORMAT_REQUEST,
  SET_FORMAT_IDS,
  RESET_FORMAT,
} from '../constants/formats';

const initialState = {
  loading: true,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

const formatsReducer = produce((draft, action = {}) => {
  switch (action.type) {
    case SET_FORMAT_LOADING:
      draft.loading = action.payload.loading;
      return;
    case ADD_FORMAT: {
      const { format } = action.payload;
      draft.items[format.id] = format;
      return;
    }
    case ADD_FORMATS: {
      const { formats } = action.payload;
      Object.assign(draft.items, formats);
      return;
    }
    case SET_FORMAT_IDS:
      draft.ids = action.payload.ids;
      return;
    case SET_FORMAT_REQUEST: {
      const { req, total } = action.payload;
      draft.req.push(req);
      draft.total = total;
      return;
    }
    case RESET_FORMAT:
      return initialState;
  }
}, initialState);

export default formatsReducer;
