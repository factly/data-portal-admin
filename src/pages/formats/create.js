import React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import FormatCreateForm from './components/FormatCreateForm';
import { createFormat } from '../../actions/formats';

const FormatCreate = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  const onCreate = (values) => {
    dispatch(createFormat(values));
    history.push('/formats');
  };

  return <FormatCreateForm onSubmit={onCreate} />;
};

export default FormatCreate;
