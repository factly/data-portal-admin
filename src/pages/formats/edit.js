import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import FormatCreateForm from './components/FormatCreateForm';
import { getFormat, updateFormat } from '../../actions/formats';

const EditFormat = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { format } = useSelector(({ formats }) => formats);

  React.useEffect(() => {
    dispatch(getFormat(id));
  }, []);

  if (!format.id) return <Skeleton />;

  const onUpdate = async (values) => {
    await dispatch(updateFormat(id, values));
    history.push('/formats');
  };

  return <FormatCreateForm data={format} onSubmit={onUpdate} />;
};

export default EditFormat;
