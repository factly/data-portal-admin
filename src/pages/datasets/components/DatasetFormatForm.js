import React from 'react';
import { notification } from 'antd';
import { useDispatch, useSelector } from 'react-redux';

import { loadFormats } from '../../../actions/formats';
import Uploader from '../../../components/Uploader';
import DatasetFormatsList from './DatasetFormatsList';

const DatasetFormatForm = ({ datasetFormats, datasetId, onSubmit, prev }) => {
  const dispatch = useDispatch();
  const formats = useSelector(({ formats }) => Object.values(formats.items));

  React.useEffect(() => {
    dispatch(loadFormats());
  }, []);

  const onFinish = (values) => {
    try {
      Object.values(values).map((file) => {
        const data = { url: file.uploadURL };
        const index = formats
          .map((format) => format.name.toLowerCase())
          .indexOf(file.extension.toLowerCase());

        if (index !== -1) {
          data.format_id = formats[index].id;
        }

        onSubmit(data)
          .then(() => {
            notification.success({
              message: 'Success',
              description: 'Dataset Format succesfully added',
            });
          })
          .catch((error) => {
            console.log(error);

            notification.error({
              message: 'Error',
              description: 'Something went wrong',
            });
          });
      });
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Uploader onUploadSuccess={onFinish} />
      <DatasetFormatsList datasetId={datasetId} datasetFormats={datasetFormats} />
    </>
  );
};

export default DatasetFormatForm;
