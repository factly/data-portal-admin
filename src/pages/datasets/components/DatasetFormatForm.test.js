import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { Button } from 'antd';
import { act } from '@testing-library/react';

import '../../../matchMedia.mock';
import DatasetFormatForm from './DatasetFormatForm';
import { loadFormats } from '../../../actions/formats';
import Uploader from '../../../components/Uploader';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../../actions/formats', () => ({
  loadFormats: jest.fn(),
}));

describe('DatasetFormatForm component', () => {
  let store;
  let mockedDispatch;

  store = mockStore({
    datasets: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
    formats: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<DatasetFormatForm />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    const datasetFormat = {
      id: 1,
      url: 'url',
      format: { name: 'format' },
      created_at: '2020-12-12',
    };
    const props = {
      datasetFormats: [datasetFormat],
      datasetId: 1,
      onSubmit: jest.fn(),
    };
    props.onSubmit.mockReturnValue(Promise.resolve());
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should render the component successfully', () => {
      useSelector.mockReturnValue({});
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <DatasetFormatForm {...props} />
          </Provider>,
        );
      });
      expect(loadFormats).toHaveBeenCalledWith();
    });
    it('should call onFinish with format', (done) => {
      useSelector.mockReturnValue([{ id: 1, name: 'pdf' }]);
      act(() => {
        wrapper = shallow(
          <Provider store={store}>
            <DatasetFormatForm {...props} />
          </Provider>,
        );
      });
      wrapper
        .find(DatasetFormatForm)
        .dive()
        .find(Uploader)
        .props()
        .onUploadSuccess([{ uploadURL: 'uploadURL', extension: 'pdf' }]);
      setTimeout(() => {
        expect(props.onSubmit).toHaveBeenCalledWith({ url: 'uploadURL', format_id: 1 });
        done();
      }, 0);
    });
    it('should call onFinish without format', (done) => {
      useSelector.mockReturnValue([{ id: 1, name: 'pdf' }]);
      act(() => {
        wrapper = shallow(
          <Provider store={store}>
            <DatasetFormatForm {...props} />
          </Provider>,
        );
      });
      wrapper
        .find(DatasetFormatForm)
        .dive()
        .find(Uploader)
        .props()
        .onUploadSuccess({ 0: { uploadURL: 'uploadURL', extension: 'csv' } });
      setTimeout(() => {
        expect(props.onSubmit).toHaveBeenCalledWith({ url: 'uploadURL' });
        done();
      }, 0);
    });
    it('should call change url', (done) => {
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = shallow(
          <Provider store={store}>
            <DatasetFormatForm {...props} />
          </Provider>,
        );
      });
      const button = wrapper.find(DatasetFormatForm).dive().find(Button);
      button.props().onClick();
      setTimeout(() => {
        expect(push).toHaveBeenCalledWith('/datasets');
        done();
      }, 0);
    });
  });
});
