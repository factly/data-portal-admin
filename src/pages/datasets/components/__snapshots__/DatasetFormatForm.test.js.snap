// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetFormatForm component snapshot testing should render the component 1`] = `
<Fragment>
  <Space>
    <Uploader
      onUploadSuccess={[Function]}
    />
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
    >
      Done
    </Button>
  </Space>
  <DatasetFormatsList />
</Fragment>
`;
